kotlinx-coroutines-core/common/src/flow/operators/Delay.kt

```diff
@@ -394,6 +394,7 @@ private fun <T> Flow<T>.timeoutInternal(
             value.onSuccess {
                 downStream.emit(it)
             }.onClosed {
+                it?.let { throw it }
```

Comment:
Is it correct that the reason this issue manifested itself is our fusing optimization?

Reply:
Oh, it boils down to the following:
```
val job = Job()
val produce = produce<Int>(job) { close(TestException()) }
(produce as Job).join()
job.isCancelled // False
```

Which makes sense for `produce` specifically (when it's used as a standalone coroutine builder), but not much when used as part of the flow machinery

Reply:
Also, kind of confusing semantics-wise: `close` with a cause **fails** the channel, and when a coroutine fails, the channel fails as well. Yet the channel failure on its now does not represent failure.